pure-wasm/src/lib.rs

```diff
@@ -191,6 +191,22 @@ pub extern "C" fn take_state(engine_ptr: u32, json_ptr: u32, json_len: u32) {
     }
 }
 
+#[unsafe(no_mangle)]
+pub extern "C" fn get_state(engine_ptr: u32) -> u32 {
+    let lock = unsafe { get_engine(engine_ptr as *const u32).unwrap() };
+    let engine = recover_lock(&lock);
+
+    let state = engine.get_state();
+    if let Ok(json_str) = serde_json::to_string(&state) {
+        let json_bytes = json_str.as_bytes();
```

Comment:
I'd rather not go through the dance of manually assembling CStrings here. It's weirdly hard to get right. Instead we can just use Rust's built in FFI handling stuff, which should work equally well for WASM. We can replace the last 3 lines of this function with

```
        let c_string = CString::new(json_bytes).unwrap();
        return c_string.into_raw() as u32
```